core/src/main/java/jenkins/slaves/WebSocketAgents.java

```diff
@@ -0,0 +1,187 @@
+        if (!(c instanceof SlaveComputer)) {
+            throw HttpResponses.notFound();
+        }
+        SlaveComputer sc = (SlaveComputer) c;
+        if (!(sc.getLauncher() instanceof JNLPLauncher)) {
+            throw HttpResponses.errorWithoutStack(400, "not an inbound agent");
+        }
+        if (!MessageDigest.isEqual(secret.getBytes(StandardCharsets.US_ASCII), sc.getJnlpMac().getBytes(StandardCharsets.US_ASCII))) { // TODO unless anonymous has CONNECT?
+            throw HttpResponses.forbidden();
+        }
+        LOGGER.fine(() -> "connecting " + agent);
+        Capability remoteCapability;
+        try (ByteArrayInputStream bais = new ByteArrayInputStream(remoteCapabilityStr.getBytes(StandardCharsets.US_ASCII))) {
+            remoteCapability = Capability.read(bais);
+            LOGGER.fine(() -> "received " + remoteCapability);
+        }
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+            new Capability().write(baos);
+            rsp.setHeader(CAPABILITY_KEY, baos.toString("US-ASCII"));
+        }
+        return WebSockets.upgrade(new Session(agent, sc, remoteCapability));
+    }
+
+    private static class Session extends WebSocketSession {
+
+        private final String agent;
+        private final SlaveComputer sc;
+        private final Capability remoteCapability;
+        private AbstractByteArrayCommandTransport.ByteArrayReceiver receiver;
+
+        Session(String agent, SlaveComputer sc, Capability remoteCapability) {
+            this.agent = agent;
+            this.sc = sc;
+            this.remoteCapability = remoteCapability;
+        }
+
+        @Override
+        protected void opened() {
+            Computer.threadPoolForRemoting.submit(() -> {
+                LOGGER.fine(() -> "setting up channel for " + agent);
+                sc.setChannel(new ChannelBuilder(agent, Computer.threadPoolForRemoting).withHeaderStream(sc.openLogFile()), new Transport(), null);
+                LOGGER.fine(() -> "set up channel for " + agent);
+                return null;
+            });
+        }
+
+        @Override
+        protected void binary(byte[] payload, int offset, int len) {
+            LOGGER.finest(() -> "reading block of length " + len + " from " + agent);
+            if (offset == 0 && len == payload.length) {
+                receiver.handle(payload);
+            } else {
+                receiver.handle(Arrays.copyOfRange(payload, offset, offset + len));
+            }
+        }
+
+        @Override
+        protected void closed(int statusCode, String reason) {
+            LOGGER.finest(() -> "closed " + statusCode + " " + reason);
+            // TODO
+        }
+
+        @Override
+        protected void error(Throwable cause) {
+            LOGGER.log(Level.WARNING, null, cause);
+        }
+
+        @Override
+        protected boolean keepAlive() {
+            return true; // Remoting ping thread may be too slow
```

Comment:
Should the Remoting ping thread be adjusted to be faster in these scenarios? Or turned off in favor of other mechanisms, like this? Duplicate ping mechanisms over a channel can sometimes get in each other's way.

I see some later comments about necessary ping intervals based on different technologies. I'm not sure what makes the most sense here.

Reply:
I think it is OK to keep both, at least pending further experience. The Remoting ping is by default every 5m with a 4m timeout, and beyond network connectivity would verify that both sides of the channel are actually processing packets. The WebSocket ping introduced here is unidirectional (server → client), does not wait for pongs unless the underlying library does it automatically, and serves only to prevent certain reverse proxies from closing the connection prematurely.